### PR TITLE
added article about time window functions

### LINKED ISSUE #105: Need document for TVF and TUMBLE window

### LINKED ISSUE #179: Document HOP window function
For example:

```
CREATE MATERIALIZED VIEW restuarant_orders AS
SELECT
    window_start,
    restaurant_id,
    COUNT(*) AS total_order
FROM 
    HOP(delivery_orders_source, order_timestamp, INTERVAL '1' MINUTE, INTERVAL '15' MINUTE)
WHERE 
    order_state = 'CREATED'
GROUP BY
    restaurant_id,
    window_start;
```

### PR DESCRIPTION
**Pull request by hengm3467**
## Info
- **Description**: 
Create an article about time window functions. It includes the concepts, how-to instructions, and examples. 

- **Related doc issues**
Resolves https://github.com/singularity-data/risingwave-docs/issues/105
Resolves https://github.com/singularity-data/risingwave-docs/issues/179

- **Related code PR**: 

- **Notes**: 
[ Any additional information? ]

## Before merging
  - [ ] (For version-specific PR) I have selected the corresponding software version in **Milestone** and linked related doc issue to this PR in **Development**.
  - [x] I have accquired the approval from the owner (and optionally the reviewers) of the code PR and at least one tech writer (`bernscode`, `CharlieSYH`, `emile-00`, & `hengm3467`). 
  - [x] I have checked the doc site preview and the updated parts look all good. <details><summary>How?</summary>image</details>
  - [ ] (For version-specific PR) I have confirmed that this PR is for a released version. If the software version is not released, put it on hold.

## Published doc pages

https://www.risingwave.dev/docs/latest/sql-function-time-window/

### CODE REVIEW

File: docs/sql/time-window-functions.md [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+In stream processing, time windows are time intervals based on which we can divide events and perform data computations.
+
+RisingWave supports two types of time windows:
+
+* Tumbling windows
+* Hopping windows
+
+For each type of time window, there is a corresponding time windowing function (hereafter referred to as “time window function”) that creates a window of this type. For tumbling windows, the function is `tumble()`. For hopping windows, the function is `hop()`.
+
+In RisingWave, the result of a time windowing function  is a table in which each row carries data for a time window. A time window function extends the schema of the original table with two new columns, `window_start` and `window_ends`, which indicate the starts and ends of time windows respectively.
```
Comment (xiangjinwu) [inline docs/sql/time-window-functions.md]: `window_end` without `s`

File: docs/sql/time-window-functions.md [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+```
+trip_id | taxi_id | completed_at | window_start | window_end 
+---------+---------+---------------------+---------------------+---------------------
+1 | 1001 | 2022-07-01 22:00:00 | 2022-07-01 22:00:00 | 2022-07-01 22 (tel:2022070122):02:00
+2 | 1002 | 2022-07-01 22:01:00 | 2022-07-01 22:00:00 | 2022-07-01 22 (tel:2022070122):02:00
+3 | 1003 | 2022-07-01 22:02:10 | 2022-07-01 22:02:00 | 2022-07-01 22 (tel:2022070122):04:00
+4 | 1004 | 2022-07-01 22:03:00 | 2022-07-01 22:02:00 | 2022-07-01 22 (tel:2022070122):04:00
+5 | 1005 | 2022-07-01 22:05:00 | 2022-07-01 22:04:00 | 2022-07-01 22 (tel:2022070122):06:00
+6 | 1006 | 2022-07-01 22:06:00 | 2022-07-01 22:06:00 | 2022-07-01 22 (tel:2022070122):08:00
```
Comment (xiangjinwu) [inline docs/sql/time-window-functions.md]: unexpected `tel:2022070122`

File: docs/sql/time-window-functions.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+GROUP BY window_start, window_end 
+ORDER BY window_start ASC;
+```
+
+The result looks like this:
+
+```
+ window_start | window_end | no_of_trips | total_distance 
+---------------------+---------------------+-------------+----------------
+ 2022-07-01 22 (tel:2022070122):00:00 | 2022-07-01 22 (tel:2022070122):02:00 | 2 | 10
```
Comment (xiangjinwu) [inline docs/sql/time-window-functions.md]: Unexpected `tel` (and more occurrences below)

File: docs/sql/time-window-functions.md [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+```
+
+### Window joins
+
+In the example below, we join two tumble time windows to get both trip and fare information.
+
+````sql
+SELECT trip.window_start, trip.window_end, trip.distance, fare.total_fare, fare.payment_status
+FROM TUMBLE (taxi_trips, completed_at, INTERVAL '2 MINUTES') as trip
+JOIN TUMBLE (taxi_fare, completed_at, INTERVAL '2 MINUTES') as fare
```
Comment (xiangjinwu) [inline docs/sql/time-window-functions.md]: The `tumble` around `taxi_fare` seems useless as neither `fare.window_start` nor `fare.window_end` is actually used anywhere.

File: docs/sql/time-window-functions.md (hunk 1 above)
Comment (hengm3467) [reply docs/sql/time-window-functions.md]: Thanks. Fixed.

File: docs/sql/time-window-functions.md (hunk 3 above)
Comment (hengm3467) [reply docs/sql/time-window-functions.md]: Thanks. Fixed all.

File: docs/sql/time-window-functions.md (hunk 4 above)
Comment (hengm3467) [reply docs/sql/time-window-functions.md]: @xiangjinwu  This is good feedback. This is more like join time window with a table. I'm (still) trying to come up with a meaningful example. Do you have suggestions? :)

File: docs/sql/time-window-functions.md (hunk 1 above)
Comment (CharlieSYH) [inline docs/sql/time-window-functions.md]: In the last paragraph, "time windowing function (hereafter referred to as “time window function”)" is already stated.
Therefore, the result of a time windowing function -> the result of a time window function .

Comment (CharlieSYH) [inline docs/sql/time-window-functions.md]: Maybe:
"which indicate the starts and ends of time windows" -> "which indicate the start and end of each time window"

File: docs/sql/time-window-functions.md [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+In RisingWave, time window functions are invoked in the **FROM** clause. See the sections below for the syntaxes of two time window functions.
+
+## `tumble()` time window function
+
+Tumbling windows are contiguous time intervals. 
+
+The syntax of the `tumble()` window function is as follows:
+
+```sql
+SELECT [ ALL | DISTINCT] [ * | *expression* [ AS *output_name* ] [, *expression* [ AS *output_name* ]...] ]
```
Comment (CharlieSYH) [inline docs/sql/time-window-functions.md]: Add an empty space after "DISTINCT"
| DISTINCT] [  ->  | DISTINCT ] [

File: docs/sql/time-window-functions.md [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+FROM TUMBLE(*table_or_source*, *start_time*, *window_size*);
+```
+
+
+*window_size* is in the format of `INTERVAL '*interval*'`. For example: `INTERVAL '2 MINUTES'`. The standard SQL format, which places time units outside of quotation marks (for example, `INTERVAL '2' MINUTE`), is also supported.
```
Comment (CharlieSYH) [inline docs/sql/time-window-functions.md]: Replace the second "for example" with "such as".

File: docs/sql/time-window-functions.md [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+|1|1001	|2022-07-01 22:00:00|4|
+|2|	1002|	2022-07-01 22:01:00	|6|
+|3|	1003|	2022-07-01 22:02:00	|3|
+|4|	1004|	2022-07-01 22:03:00	|7|
+|5|	1005|	2022-07-01 22:05:00	|2|
+|6|	1006|	2022-07-01 22:05:30	|8|
+
+
+
+Here is an example that uses tumble window function.
```
Comment (CharlieSYH) [inline docs/sql/time-window-functions.md]: Add "the" before "tumble window function".

File: docs/sql/time-window-functions.md [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+
+
+## `hop()` time window function
+
+Hopping windows are scheduled time intervals. A hopping window consists of three time-related parameters: start time, hop size, and window size.
+
+See below for the syntax of the `hop()` window function.
+
+```sql
+SELECT [ ALL | DISTINCT] [ * | expression [ AS output_name ] [, expression [ AS output_name ]...] ]
```
Comment (CharlieSYH) [inline docs/sql/time-window-functions.md]: Add a space after DISTINCT

File: docs/sql/time-window-functions.md [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+
+The syntax of the `tumble()` window function is as follows:
+
+```sql
+SELECT [ ALL | DISTINCT] [ * | expression [ AS output_name ] [, expression [ AS output_name ]...] ]
+FROM TUMBLE(table_or_source, start_time, window_size);
+```
+
+
+*window_size* is in the format of `INTERVAL '<interval>'`. For example: `INTERVAL '2 MINUTES'`. The standard SQL format, which places time units outside of quotation marks (for example, `INTERVAL '2' MINUTE`), is also supported.
```
Comment (CharlieSYH) [inline docs/sql/time-window-functions.md]: '< >' adds confusion when we use it to wrap up the struct data type.
Suggest replacing `INTERVAL '<interval>'` with <code>INTERVAL '<i>interval</i>'</code>

File: docs/sql/time-window-functions.md [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+Hopping windows are scheduled time intervals. A hopping window consists of three time-related parameters: start time, hop size, and window size.
+
+See below for the syntax of the `hop()` window function.
+
+```sql
+SELECT [ ALL | DISTINCT] [ * | expression [ AS output_name ] [, expression [ AS output_name ]...] ]
+FROM HOP(table_or_source, start_time, hop_size, window_size);
+```
+
+Both *hop_size* and *window_size* are in the format of `INTERVAL '<interval>'`. For example: `INTERVAL '2 MINUTES'`. The standard SQL format, which places time units outside of quotation marks (for example, `INTERVAL '2' MINUTE`), is also supported.
```
Comment (CharlieSYH) [inline docs/sql/time-window-functions.md]: Suggest replacing `INTERVAL '<interval>'` with <code>INTERVAL '<i>interval</i>'</code>

File: docs/sql/time-window-functions.md (hunk 4 above)
Comment (hengm3467) [reply docs/sql/time-window-functions.md]: I added a join condition "trip.window_start = fare.window_start". At least the time attributes are being used. Hope this works for now.

File: docs/sql/time-window-functions.md [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+
+Here is an example.
+
+```sql
+SELECT trip_id, taxi_id, completed_at, window_start, window_end
+FROM HOP (taxi_trips, completed_at, INTERVAL '1 MINUTE', INTERVAL '2 MINUTES')
```
Comment (CharlieSYH) [inline docs/sql/time-window-functions.md]: Maybe we can use standard and non-standard interval arguments together:
`FROM HOP (taxi_trips, completed_at, INTERVAL '1 MINUTE', INTERVAL '2‘ MINUTE)`

File: docs/sql/time-window-functions.md [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+ 2022-07-01 22:02:00 | 2022-07-01 22:04:00 | 7        | BEST TAXI
+ 2022-07-01 22:04:00 | 2022-07-01 22:06:00 | 2        | WEST TAXI
+ 2022-07-01 22:06:00 | 2022-07-01 22:08:00 | 8        | EAST TAXI
+```
+
+### Window joins
+
+In the example below, we join two tumble time windows to get both trip and fare information.
+
+````sql
```
Comment (CharlieSYH) [inline docs/sql/time-window-functions.md]: typo:
````sql  -> ```sql

File: docs/sql/time-window-functions.md (hunk 6 above)
Comment (hengm3467) [reply docs/sql/time-window-functions.md]: I'll change it to:
Example: `...`.

File: docs/sql/time-window-functions.md (hunk 4 above)
Comment (hengm3467) [reply docs/sql/time-window-functions.md]: @xiangjinwu Do you think the query is OK for now? or we need to come up with a new one? Thanks. :)

Comment (xiangjinwu) [reply docs/sql/time-window-functions.md]: I think it is okay now as it reflects the general pattern that windows of 2 tables should match.

Comment (hengm3467) [reply docs/sql/time-window-functions.md]: Cool. Thanks!

File: docs/sql/time-window-functions.md (hunk 11 above)
Comment (hengm3467) [reply docs/sql/time-window-functions.md]: That's not a recommended practice. :) Let's keep it the way it is.

**Review by CharlieSYH — APPROVED** [review]

File: docs/sql/functions-operators/sql-function-time-window.md [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+In stream processing, time windows are time intervals based on which we can divide events and perform data computations.
+
+RisingWave supports two types of time windows:
+
+* Tumbling windows
+* Hopping windows
+
+For each type of time window, there is a corresponding time windowing function (hereafter referred to as “time window function”) that creates a window of this type. For tumbling windows, the function is `tumble()`. For hopping windows, the function is `hop()`.
+
+In RisingWave, the result of a time window function  is a table in which each row carries data for a time window. A time window function extends the schema of the original table with two new columns, `window_start` and `window_end`, which indicate the start and end of time windows respectively.
```
Comment (TennyZhuang) [inline docs/sql/functions-operators/sql-function-time-window.md]:
```suggestion
In RisingWave, the result of a time window function is a table in which each row carries data for a time window. A time window function extends the schema of the original table with two new columns, `window_start` and `window_end`, which indicate the start and end of time windows respectively.
```

**Review by TennyZhuang — APPROVED** [review]
LGTM!

**Pull request merged**